### PR TITLE
Revise modular arithmetic using Periods

### PR DESCRIPTION
**Pull request by omus**
After playing around with modular arithmetic with `Period`s I discovered that `mod` and `rem` had some strange behaviours:

``` julia
julia> using Base.Dates

julia> x, y = Second(5), Second(2)
(5 seconds,2 seconds)

julia> div(x,y)  # Correct
2

julia> rem(x,y)  # Expected result to be `Second(1)`
1

julia> div(x,y)*y + rem(x,y) == x  # Should return `true`
ERROR: MethodError: no method matching +(::Base.Dates.Second, ::Int64)
Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...)
  +(::Char, ::Integer)
  +(::Complex{Bool}, ::Real)
  ...
 in eval(::Module, ::Any) at ./boot.jl:236
```

I have updated the code such that `mod` and `rem` now return `Period`s rather than `Int64`s. Additionally I also added support for `fld` for `Period`s.

### CODE REVIEW

**Comment by omus** [comment]
cc: @quinnj

**Comment by quinnj** [comment]
@stevengj did a big cleanup of the Period arithmetic operations at one point too.

**Comment by stevengj** [comment]
Yup, looks like this was my fault: I messed up the remainder case in #9220.

File: base/dates/periods.jl [hunk 1]
```diff
@@ -65,27 +65,34 @@ Base.isless{P<:Period}(x::P,y::P) = isless(value(x),value(y))
...
+for op in (:/,:div,:fld)
     @eval begin
         ($op){P<:Period}(x::P,y::P) = ($op)(value(x),value(y))
         ($op){P<:Period}(x::P,y::Real) = P(($op)(value(x),Int64(y)))
     end
 end
+
+for op in (:rem,:mod)
+    @eval begin
+        ($op){P<:Period}(x::P,y::P) = P(($op)(value(x),value(y)))
```
Comment (stevengj) [inline base/dates/periods.jl]: Shouldn't we also have `($op){P<:Period}(x::P,y::Real) = P(($op)(value(x),y))`?

Comment (omus) [inline base/dates/periods.jl]: I intentionally left out that method definition as I was unsure whether that was actually a real use case for `rem(Second(5), 2)`. I think I'd prefer to be cautious and leave it out for now.

Comment (stevengj) [inline base/dates/periods.jl]: Since we define `div(Second(5), 2)`, we should also define `rem`.

**Comment by stevengj** [comment]
LGTM.

**Pull request merged**